Avoid permanent change of the IdGenerator entity metadata

soullivaneuh opened the pull request:
fixes #105

In any case the generator is changed on the metadata, it must be restored afterward.

Otherwise, manual entity persistence after fixture loading will cause troubles.

Tests was added during the feature addition: https://github.com/theofidry/AliceDataFixtures/commit/7053c75de4d5a91c5bd7db43918a907706c7c990#diff-292d5068b7f30308701962493b4cfd57

So if Travis is green, that will prove the condition is not necessary.

Otherwise, we have to find an another way to deal with it, because leaving the metadata at this state is not acceptable.

soullivaneuh commented:
> So if Travis is green

Well, looks like Travis does not give a single f*ck of my PR.

theofidry commented:
>Otherwise, we have to find an another way to deal with it, because leaving the metadata at this state is not acceptable.

I'm not 100% sure to understand how the generator metadata works to be honest: if its state matters before the persist or if it does until the flush. If the later, we can restore it in the `flush()` instead

For the tests I don't know why Travis is not being triggered... Did you try `make test` locally?

soullivaneuh commented:
> If the later, we can restore it in the flush() instead

We don't have entity access on flush event. Are you suggesting to store them on the class to reset them later? That's a possibility, but are we sure we need this? What the condition I removed does exactly?

> For the tests I don't know why Travis is not being triggered... Did you try make test locally?

The `make test` require a lot of system dependencies to install, it's quite hard to make it running locally.

Do you have access to the Travis admin panel of this project? Maybe the integration is broken.

theofidry commented:
>Do you have access to the Travis admin panel of this project? Maybe the integration is broken.

There's the webhook and it seems to be working fine. That said I saw Travis behaving weirdly those last few weeks so it might have some trouble right now...

>We don't have entity access on flush event. Are you suggesting to store them on the class to reset them later? That's a possibility, but are we sure we need this? What the condition I removed does exactly?

I'm not sure what the condition you removed does exactly. But yeah we could store the necessary information to restore the metadata later on. I don't have a strong opinion about which solution should be used: I think it mostly depends on whether or not the metadata is used between the persist and the flush

soullivaneuh commented:
I propose to let you find why Travis is not working anymore first.

Then if test are green, just use this solution. If this lead to another errors we will be able to correctly identify them and add comments with the condition, and we will use the flush method.

What do you think?

theofidry commented:
Ok let's do that

soullivaneuh commented:
Travis job looks launched: https://travis-ci.org/github/theofidry/AliceDataFixtures/builds/683853102

Maybe a configuration issue with the GitHub repository? Try to re-enable the project on Travis, it may fix the problem.

theofidry commented:
Double checked and re-enabled

soullivaneuh commented:
Tests are failing. Not really sure what @Nijusan wanted to do with this condition (maybe he will respond!), but I maybe find a cleaner solution to restore all at flush.

soullivaneuh commented on src/Bridge/Doctrine/Persister/ObjectManagerPersister.php:
```diff
@@ -123,19 +116,23 @@ private function getPersistableClasses(ObjectManager $manager): array
+        }
     }
 
-    private function getMetadata(string $class): ClassMetadata
```
This metadata cache is useless has doctrine already has one: https://github.com/doctrine/orm/blob/9a8e1fb47776fc5ff49f995f3fef93a5ff8469b0/lib/Doctrine/ORM/Mapping/ClassMetadataFactory.php#L239-L241

Removed.

soullivaneuh commented on src/Bridge/Doctrine/Persister/ObjectManagerPersister.php:
```diff
@@ -123,19 +116,23 @@ private function getPersistableClasses(ObjectManager $manager): array
     }
 
-    private function getMetadata(string $class): ClassMetadata
+    private function configureIdGenerator(ORMClassMetadataInfo $metadata): ORMClassMetadataInfo
```
The `configureIdGenerator` has no reason to be `protected` AFAIK, so I make it private and changed its signature.

soullivaneuh commented:
And voilà! Did a clone of the metadata and replaced it with the new one.

Then, restored and flush.

Also simplify a bit the codebase and added a proof test.

I leave some explanations on the diff. Ready for review. :+1:

soullivaneuh commented:
Hmm looks like travis jobs are not triggered at all now: https://travis-ci.org/github/theofidry/AliceDataFixtures/pull_requests

And the last job of #136 is failing. So failing on this one may not be related.

I may take a look to the settings with granted access if you want to.

soullivaneuh commented:
Ran `make test_doctrine_bridge` successfully on my local.

theofidry commented:
Will take a fresh look tomorrow

Nijusan commented:
> Tests are failing. Not really sure what @Nijusan wanted to do with this condition (maybe he will respond!), but I maybe found a cleaner solution to restore all at flush.

hey @soullivaneuh , sorry for late response, which condition do u mean? long time ago i worked on this projekt :)

theofidry commented:
@soullivaneuh could you try to push an empty commit now to see if that works?

theofidry commented:
@soullivaneuh Travis should be fixed and Symfony 5 support has been merged

Nijusan commented:
> Tests are failing. Not really sure what @Nijusan wanted to do with this condition (maybe he will respond!), but I maybe found a cleaner solution to restore all at flush.

maybe it's still useful information: i recently updated a project from symfony 4.4 to 5.1 and ran into problems after loading my fixtures. and i remembered why i added this condition: it was not possible to mix fixtures for one entity some with given id-information and some relying on a post insert generator. so it has to be all or nothing. either all fixtures of this entity have  the id set in fixtures or none of them. 

this solution worked fine until the update. now problems occur in a specific phpunit test, where first fixtures of one entity are loaded with given id information, and in the test another new entity is tried to persist and fails then with wrong parameter exception...

theofidry commented:
@Nijusan I'm not sure to understand the solution you're proposing here: you mean the change to the metadata should be done right before the flush (right now it's done during the first pass) and restored after (like done in this PR)? Or are you suggesting to use another hook?

Nijusan commented:
@theofidry sorry for the confusion. i just wanted to point out why i implemented this check. given an entity with primary key id, it was not possible to have a fixture file with (in the case of a post insert id generator):

```
Entity\MyEntity:
    my_entity_1:
        id: 3
       name: test

    my_entity_2:
        name: foo
```
 if i restore the metadata in this case before the flush, the given id: 3 from my fixture wont be used. and because of that, we said if we want to manual define the primary key, all given entities have to do so.

so i think the right thing is to restore the metadata after the flush (like done in this PR), still having the problem with all or nothing (all entites have explicit primary kes in fixture file or none of them). but i tried the changes in my project without success, not all fixtures were loaded correctly :( I'll try to take some time to check why, maybe it was just a mistake of mine

theofidry commented:
Thank you @soullivaneuh

theofidry commented:
@Nijusan yes it's an all or nothing right now, but this is more a Doctrine limitation than alice unfortunately

yakobe commented:
Any chance we can get this released? 🙂

theofidry commented:
done

yakobe commented:
Thank you!

tristanbes commented:
I can confirm that this PR breaks previously working fixtures with error:

```
No ID found for the entity "App\Entity\Config". If this entity has an auto ID generator, this may be due to having it disabled because one instance of the entity had an ID assigned. Either remove this assigned ID to allow the auto ID generator to operate or generate and ID for all the "App\Entity\Config" entities
```

For the given fixture:

```yml
App\Entity\Config:
    config_etude_lyon_fidgar:
        name: 'Fid Garage'
        codeService: 'FidGar'
        roles: []
        __calls:
            - setType: [<(App\Entity\Config::TYPE_DEM)>]
            - addTiers: ['@tiers_with_another_config']
```

theofidry commented:
A wild guess: the metadata is injected & cached somewhere resulting in the current restore not to work. I'll revert this PR and release a patch. We'll just go forward and deprecate this feature and remove it in the next major version.

I'm a bit under water atm so if I forget it this weekend don't hesitate to ping me next week

yakobe commented:
Any news on this? Would be good to get a patch.

We are currently stuck on https://github.com/theofidry/alicedatafixtures/releases/tag/1.0.1 because the changes to id, but we need https://github.com/theofidry/AliceDataFixtures/releases/tag/v1.1.2 to use symfony 5 because of https://github.com/theofidry/AliceDataFixtures/issues/123 😛 

Maybe it's possible to revert all the breaking changes to the id since 1.0.1 😉😬

theofidry commented:
>Maybe it's possible to revert all the breaking changes to the id since 1.0.1 😉😬

As far as I'm aware there is only one which is this PR. I'll gladly merge & tag a release reverting it

yakobe commented:
The problem is that this PR fixes https://github.com/theofidry/AliceDataFixtures/issues/105. That means if it is rolled back then we still have the problem in https://github.com/theofidry/AliceDataFixtures/issues/105.

theofidry commented:
But the problem described in #105 was existing from day one and I think the real fix is removing this hybrid approach.

Maybe though as an alternative to completely reverting #105 the restore could be done via a toggle, which would allow some like @soullivaneuh to still benefit from it; just not enabled by default since it appears to easily break things

yakobe commented:
I think we might be misunderstanding eachother a little 😄.

The bug in #105 was introduced in https://github.com/theofidry/alicedatafixtures/releases/tag/1.1.0. 
This pull request (#135) attempts to fix it but introduces a different bug. If we revert this pull request we still have the bug described in #105. In order to fix #105 we would need to revert the original code added in https://github.com/theofidry/alicedatafixtures/releases/tag/1.1.0 that caused it.

Maybe the best solution is to just take away the new feature of being able to manually assign id's that are auto-generated by a database and not do any manipulation of the schema. That would be the preference since it can very easily cause nasty side-effects. Let the database generate the id as it is meant to. We have other ways to handle id references if we need 🙂

theofidry commented:
>Maybe the best solution is to just take away the new feature of being able to manually assign id's that are auto-generated by a database and not do any manipulation of the schema.

Just to be clear: this is not a new feature, this is from day 1 because was already a feature in Alice 2.x.

It's a chicken egg problem, #105 fixes an issue of the metadata not being restored but yet restoring it after a flush is creating a different set of issues. Hence IMO:

- short term; make the fix introduced in #105 toggable (but is not the default)
- long term: deprecate the feature and remove it in 2.x

yakobe commented:
No, sorry. #105 is an issue that reports a bug introduced on v1.1.0. 

#135 is the PR that tries to fix it but introduces a new bug.
